packages/ui/aurora/src/components/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,200 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { CaretDown, CaretUp } from '@phosphor-icons/react';
+import { createContextScope, Scope } from '@radix-ui/react-context';
+import { createPopperScope } from '@radix-ui/react-popper';
+import * as PopperPrimitive from '@radix-ui/react-popper';
+import { useCombobox, UseComboboxReturnValue } from 'downshift';
+import React, { forwardRef, PropsWithChildren } from 'react';
+
+import { useThemeContext } from '../../hooks';
+import { ThemedClassName } from '../../util';
+import { Button } from '../Buttons';
+import { Input } from '../Input';
+
+const COMBOBOX_NAME = 'ComboBox';
+
+type ScopedProps<P> = P & { __scopeComboBox?: Scope };
+
+type ValueAdapter<T> = (value: T) => { id: string; text: string };
+
+type ComboBoxContextValue<T> = {
+  items: T[];
+  adapter: ValueAdapter<T>;
+} & UseComboboxReturnValue<T>;
+
+const usePopperScope = createPopperScope();
+const [createComboBoxContext] = createContextScope(COMBOBOX_NAME, [createPopperScope]);
+const [ComboBoxProvider, useComboBoxContext] = createComboBoxContext<ComboBoxContextValue<any>>(COMBOBOX_NAME);
+
+type ComboBoxRootProps<T> = ThemedClassName<
+  ScopedProps<
+    PropsWithChildren<{
+      placeholder?: string;
+      items?: T[];
+      value?: T;
+      adapter: ValueAdapter<T>;
+      onChange?: (value: T | undefined) => void;
+      onInputChange?: (text?: string) => void;
+    }>
+  >
+>;
+
+/**
+ * Type-ahead combobox.
+ * https://www.downshift-js.com
+ */
+const ComboBoxRoot = <T,>({
```

Comment:
I haven’t seen generics used this way in the Radix ecosystem. I'm not sure we can expect developers will make use of this, and it could cause trouble if it has no defaults.

Reply:
Maybe it becomes irrelevant with subcomponents, since this appears to govern the schema of `items`, which should instead be the props of some `ComboboxItem` component.

Reply:
It's actually needed for the ValueAdapter.